README.md
[![verified-by-homebridge](https://badgen.net/badge/homebridge/verified/purple)](https://github.com/homebridge/homebridge/wiki/Verified-Plugins)
[![npm version](https://badge.fury.io/js/homebridge-3em-energy-meter.svg)](https://www.npmjs.com/package/homebridge-3em-energy-meter)
[![Downloads](https://img.shields.io/npm/dt/homebridge-3em-energy-meter.svg)](https://www.npmjs.com/package/homebridge-3em-energy-meter)

# Homebridge 3em Energy Meter

[Homebridge 3em Energy Meter](https://www.npmjs.com/package/homebridge-3em-energy-meter) is a plugin for [Homebridge](https://github.com/homebridge/homebridge) that implements Shelly 3EM energy metering functionality in Homekit.

This plugin uses http requests to a Shelly 3EM (or EM*) device, making it possible to retain the native Shelly cloud statistics (which use MQTT) and at the same time allow you to monitor your energy consumption via Homekit. 

Please note that due to the fact that Apple does not support energy characteristics in Homekit, this plugin's accessory will only show values in the third-party homekit application "EVE".

[![Status](screenshots/homebridge-3em-energy-meter-eve-app.png)]

It will show in the EVE application the following values: Voltage (the average voltage of all 3 phases), Current (the accumulated Ampere of all 3 phases), Consumption (the accumulated Watts of all 3 phases) and the Total Consumption (the accumulated kWh of all 3 phases as calculated by Shelly API. Note in order to reset this value you must reset it in the Shelly app). A Total Cost and Projected Cost will show if you have specified the Energy Cost in the settings section of your EVE application. Total Consumption and Total Cost will feature the fakegato-history graph.

This application uses the cool fakegato plugin ([simont77/fakegato-history](https://github.com/simont77/fakegato-history)).

Please also note that the sole purpose of this plugin is the energy metering feature of the Shelly 3em in order to monitor your 3 phase installation. The Shelly 3em features also an actuator (switch) which is not implemented in this plugin. If you need to use the switch in Homekit please use other plugins, like for example the very good homebridge-shelly ([alexryd/homebridge-shelly](https://github.com/alexryd/homebridge-shelly)) plugin.

# Installation Instructions

You can easily install this plugin by using the superb [Homebridge Config UI X](https://www.npmjs.com/package/homebridge-config-ui-x). Search for "homebridge-3em-energy-meter" in the "Plugins" tab and install the plugin. After that fill out the configuration by using the "SETTINGS" link below the installed plugin.

Alternatively, can you install the plugin by 

             npm install -g homebridge-3em-energy-meter

and then edit your Homebridge's config.json to include the following in the accessories section:

        {
            "accessory": "3EMEnergyMeter",
            "name": "Energy Meter",
            "ip": "192.168.0.1",
            "auth": {
                "user": "",
                "pass": ""
            },
            "timeout": 5000,
            "update_interval": 10000,
            "use_em": false,
            "use_em_mode": 0,
            "negative_handling_mode": 0,
            "use_pf": false,
            "debug_log": false,
            "serial": "123456789012345"             
        },

* "name"              			The Homekit Accessory Name.
* "ip"                			The IP address of your Shelly 3EM.
* "user" and "pass"   			If your Shelly 3EM local web page is password protected specify "user" and "pass".
* "timeout"           			The http/get request timeout in msec. This timeout must the less than the "update_interval", default is 5000.
* "update_interval"   			The interval for pulling values in msec. Must be greater than "timeout" setting, default is 10000.
* "use_em"            			Use this plugin with a Shelly EM device.
* "use_em_mode" 						Set the mode when use_em is true. Set to 0 to combine channel1 and channel2. Use 1,2 when single channels should be used,respectively.
* "use_pf"            			Enables the Power Factor (pf) usage when calculating Total Ampere.
* "negative_handling_mode"	Defines what happens with negative values. Set to 0 to zero them or 1 to show them as absolute values.
* "debug_log"         			Enables the debug logging of the plugin, default is false.
* "serial"            			This sets the published serialNumber of the accessory. It is required to use an unique serial for fakegato-history to work correctly.

Shelly EM functionality is beta, please use at own risk as I could not test it on a real EM device.
The creator of this plugin is not affiliated in any way with [Shelly(Allterco)](https://shelly.cloud/) or [EVE](https://www.evehome.com/).

[![Support via PayPal](https://cdn.rawgit.com/twolfson/paypal-github-button/1.0.0/dist/button.svg)](https://www.paypal.me/produdegr/)


CHANGELOG.md
# Change Log

## 1.1.3 (2021-05-18)

### Changes

* Added a mode selection in order to specify what to do when negative values appear (Power returns etc.).


## 1.1.2 (2021-11-02)

### Changes

* Added correct absolute ( abs() ) to calculations in order to comply to Homekit ranges (no negative values allowed).

## 1.1.1 (2021-11-02)

### Changes

* Added absolute ( abs() ) to calculations in order to comply to Homekit ranges (no negative values allowed).

## 1.1.0 (2021-08-02)

### Changes

* Added support for Shelly EM devices (beta)
* Please set config flag use_em to true and 
  use use_em_mode to get combined, channel1 or channel2 (setting 0,1,2)
  to use this plugin with a Shelly EM.

## 1.0.0 (2021-06-11)

### Changes

* Bumped stable and tested release to major version 1.0.0
* Just added donation button ;)

## 0.1.3 (2021-04-21)

### Changes

* Added option to use the Power Factor (pf) when calculating Total Ampere.


## 0.1.2 (2021-04-10)

### Changes

* Added returned metered values to debug log.



index.js
var inherits = require('util').inherits;
var Service, Characteristic;
var request = require('request');
var FakeGatoHistoryService = require('fakegato-history');
const version = require('./package.json').version;

module.exports = function (homebridge) {
	Service = homebridge.hap.Service;
	Characteristic = homebridge.hap.Characteristic;
	Accessory = homebridge.platformAccessory;
	UUIDGen = homebridge.hap.uuid;
	FakeGatoHistoryService = require('fakegato-history')(homebridge);
	homebridge.registerAccessory("homebridge-3em-energy-meter", "3EMEnergyMeter", EnergyMeter);
}

function EnergyMeter (log, config) {
	this.log = log;
	this.ip = config["ip"] || "127.0.0.1";
	this.url = "http://" + this.ip + "/status/emeters?";
	this.auth = config["auth"];
	this.name = config["name"];
	this.displayName = config["name"];
	this.timeout = config["timeout"] || 5000;
	this.http_method = "GET";
	this.update_interval = Number(config["update_interval"] || 10000);
	this.use_em = config["use_em"] || false;
	this.use_em_mode = config["use_em_mode"] || 0; 
	this.negative_handling_mode = config["negative_handling_mode"] || 0; 	
	this.use_pf = config["use_pf"] || false;
	this.debug_log = config["debug_log"] || false;
	this.serial = config.serial || "9000000";

	// internal variables
	this.waiting_response = false;
	this.powerConsumption = 0;
	this.totalPowerConsumption = 0;
	this.voltage1 = 0;
	this.ampere1 = 0;
	this.pf0 = 1;
	this.pf1 = 1;
	this.pf2 = 1;

	var EvePowerConsumption = function () {
		Characteristic.call(this, 'Consumption', 'E863F10D-079E-48FF-8F27-9C2605A29F52');
		this.setProps({
			format: Characteristic.Formats.UINT16,
			unit: "Watts",
			maxValue: 100000,
			minValue: 0,
			minStep: 1,
			perms: [Characteristic.Perms.READ, Characteristic.Perms.NOTIFY]
		});
		this.value = this.getDefaultValue();
	};
	EvePowerConsumption.UUID = 'E863F10D-079E-48FF-8F27-9C2605A29F52';
	inherits(EvePowerConsumption, Characteristic);

	var EveTotalConsumption = function () {
		Characteristic.call(this, 'Energy', 'E863F10C-079E-48FF-8F27-9C2605A29F52');
		this.setProps({
			format: Characteristic.Formats.FLOAT,
			unit: 'kWh',
			maxValue: 1000000000,
			minValue: 0,
			minStep: 0.001,
			perms: [Characteristic.Perms.READ, Characteristic.Perms.NOTIFY]
		});
		this.value = this.getDefaultValue();
	};
	EveTotalConsumption.UUID = 'E863F10C-079E-48FF-8F27-9C2605A29F52';
	inherits(EveTotalConsumption, Characteristic);

	var EveVoltage1 = function () {
		Characteristic.call(this, 'Volt', 'E863F10A-079E-48FF-8F27-9C2605A29F52');
		this.setProps({
			format: Characteristic.Formats.FLOAT,
			unit: 'Volt',
			maxValue: 1000000000,
			minValue: 0,
			minStep: 0.001,
			perms: [Characteristic.Perms.READ, Characteristic.Perms.NOTIFY]
		});
		this.value = this.getDefaultValue();
	};
	EveVoltage1.UUID = 'E863F10A-079E-48FF-8F27-9C2605A29F52';
	inherits(EveVoltage1, Characteristic);

	var EveAmpere1 = function () {
		Characteristic.call(this, 'Ampere', 'E863F126-079E-48FF-8F27-9C2605A29F52');
		this.setProps({
			format: Characteristic.Formats.FLOAT,
			unit: 'Ampere',
			maxValue: 1000000000,
			minValue: 0,
			minStep: 0.001,
			perms: [Characteristic.Perms.READ, Characteristic.Perms.NOTIFY]
		});
		this.value = this.getDefaultValue();
	};
	EveAmpere1.UUID = 'E863F126-079E-48FF-8F27-9C2605A29F52';
	inherits(EveAmpere1, Characteristic);

	var PowerMeterService = function (displayName, subtype) {
		Service.call(this, displayName, '00000001-0000-1777-8000-775D67EC4377', subtype);
		this.addCharacteristic(EvePowerConsumption);
		this.addOptionalCharacteristic(EveTotalConsumption);
		this.addOptionalCharacteristic(EveVoltage1);
    this.addOptionalCharacteristic(EveAmpere1);
	};
	PowerMeterService.UUID = '00000001-0000-1777-8000-775D67EC4377';
	inherits(PowerMeterService, Service);

	// local vars
	this._EvePowerConsumption = EvePowerConsumption;
	this._EveTotalConsumption = EveTotalConsumption;
	this._EveVoltage1 = EveVoltage1;
  this._EveAmpere1 = EveAmpere1;

  // info
  this.informationService = new Service.AccessoryInformation();
	this.informationService
			.setCharacteristic(Characteristic.Manufacturer, "Shelly - produdegr")
			.setCharacteristic(Characteristic.Model, "Shelly 3EM")
			.setCharacteristic(Characteristic.FirmwareRevision, version)
			.setCharacteristic(Characteristic.SerialNumber, this.serial);

	// construct service
	this.service = new PowerMeterService(this.name);
	this.service.getCharacteristic(this._EvePowerConsumption).on('get', this.getPowerConsumption.bind(this));
	this.service.addCharacteristic(this._EveTotalConsumption).on('get', this.getTotalConsumption.bind(this));
  this.service.addCharacteristic(this._EveVoltage1).on('get', this.getVoltage1.bind(this));
  this.service.addCharacteristic(this._EveAmpere1).on('get', this.getAmpere1.bind(this));

  // add fakegato
  this.historyService = new FakeGatoHistoryService("energy", this,{storage:'fs'});
}

EnergyMeter.prototype.updateState = function () {
	if (this.waiting_response) {
		this.log('Please select a higher update_interval value. Http command may not finish!');
		return;
	}
	this.waiting_response = true;
	this.last_value = new Promise((resolve, reject) => {
		var ops = {
			uri:		this.url,
			method:		this.http_method,
			timeout:	this.timeout
		};
		if (this.debug_log) { this.log('Requesting energy values from Shelly 3EM(EM) ...'); }
		if (this.auth) {
			ops.auth = {
				user: this.auth.user,
				pass: this.auth.pass
			};
		}
		request(ops, (error, res, body) => {
			var json = null;
			if (error) {
				this.log('Bad http response! (' + ops.uri + '): ' + error.message);
			}
			else {
				try {
					json = JSON.parse(body);
					
					if ((this.use_pf) && (this.use_em==false)) {
						this.pf0 = parseFloat(json.emeters[0].pf);
						this.pf1 = parseFloat(json.emeters[1].pf);
						this.pf2 = parseFloat(json.emeters[2].pf);
					}
					else {
						this.pf0 = 1;
						this.pf1 = 1;
						this.pf2 = 1;
					}
					
					if (this.use_em) {
						
						if (this.use_em_mode == 0) {
					   if (this.negative_handling_mode == 0) {
					   	this.powerConsumption = (parseFloat(json.emeters[0].power)+parseFloat(json.emeters[1].power));
					    this.totalPowerConsumption = ((parseFloat(json.emeters[0].total)+parseFloat(json.emeters[1].total))/1000);
					    this.voltage1 = (((parseFloat(json.emeters[0].voltage)+parseFloat(json.emeters[1].voltage))/2));
				    	this.ampere1 = ((this.powerConsumption/this.voltage1));
				    	if (this.powerConsumption < 0) { this.powerConsumption = 0 }
				    	if (this.totalPowerConsumption < 0) { this.totalPowerConsumption = 0 }
				    	if (this.voltage1 < 0) { this.voltage1 = 0 }
				    	if (this.ampere1 < 0) { this.ampere1 = 0 }
					  } else if (this.negative_handling_mode == 1) {
					    this.powerConsumption = Math.abs(parseFloat(json.emeters[0].power)+parseFloat(json.emeters[1].power));
					    this.totalPowerConsumption = Math.abs((parseFloat(json.emeters[0].total)+parseFloat(json.emeters[1].total))/1000);
					    this.voltage1 = Math.abs(((parseFloat(json.emeters[0].voltage)+parseFloat(json.emeters[1].voltage))/2));
				    	    this.ampere1 = Math.abs((this.powerConsumption/this.voltage1));
				    	}
						  	
					  } else
					  	{ if (this.use_em_mode == 1) {
					        if (this.negative_handling_mode == 0) {
					        this.powerConsumption = (parseFloat(json.emeters[0].power));
					        this.totalPowerConsumption = (parseFloat(json.emeters[0].total)/1000);
					        this.voltage1 = (parseFloat(json.emeters[0].voltage));
				        	this.ampere1 = ((this.powerConsumption/this.voltage1));
				    	    if (this.powerConsumption < 0) { this.powerConsumption = 0 }
				    	    if (this.totalPowerConsumption < 0) { this.totalPowerConsumption = 0 }
				    	    if (this.voltage1 < 0) { this.voltage1 = 0 }
				    	    if (this.ampere1 < 0) { this.ampere1 = 0 }
					        } else if (this.negative_handling_mode == 1) {
					        this.powerConsumption = Math.abs(parseFloat(json.emeters[0].power));
					        this.totalPowerConsumption = Math.abs(parseFloat(json.emeters[0].total)/1000);
					        this.voltage1 = Math.abs(parseFloat(json.emeters[0].voltage));
				        	this.ampere1 = Math.abs((this.powerConsumption/this.voltage1));
				      }
						  	
					    } else 
					      { if (this.use_em_mode == 2) {
					        if (this.negative_handling_mode == 0) {
					        this.powerConsumption = (parseFloat(json.emeters[1].power));
					        this.totalPowerConsumption = (parseFloat(json.emeters[1].total)/1000);
					        this.voltage1 = (parseFloat(json.emeters[1].voltage));
				          this.ampere1 = ((this.powerConsumption/this.voltage1));
				    	    if (this.powerConsumption < 0) { this.powerConsumption = 0 }
				    	    if (this.totalPowerConsumption < 0) { this.totalPowerConsumption = 0 }
				    	    if (this.voltage1 < 0) { this.voltage1 = 0 }
				    	    if (this.ampere1 < 0) { this.ampere1 = 0 }
					        } else if (this.negative_handling_mode == 1) {
					          this.powerConsumption = Math.abs(parseFloat(json.emeters[1].power));
					          this.totalPowerConsumption = Math.abs(parseFloat(json.emeters[1].total)/1000);
					          this.voltage1 = Math.abs(parseFloat(json.emeters[1].voltage));
				          	this.ampere1 = Math.abs((this.powerConsumption/this.voltage1));
				          	}
						  	
					        }
					        }	
					  	} 
						} 
						else {
						 			if (this.negative_handling_mode == 0) {
				           this.powerConsumption = (parseFloat(json.emeters[0].power)+parseFloat(json.emeters[1].power)+parseFloat(json.emeters[2].power));
				           this.totalPowerConsumption = ((parseFloat(json.emeters[0].total)+parseFloat(json.emeters[1].total)+parseFloat(json.emeters[2].total))/1000);
			             this.voltage1 = (((parseFloat(json.emeters[0].voltage)+parseFloat(json.emeters[1].voltage)+parseFloat(json.emeters[2].voltage))/3));
				           this.ampere1 = (((parseFloat(json.emeters[0].current)*this.pf0)
					          +(parseFloat(json.emeters[1].current)*this.pf1)
					          +(parseFloat(json.emeters[2].current)*this.pf2)));
				    	    if (this.powerConsumption < 0) { this.powerConsumption = 0 }
				    	    if (this.totalPowerConsumption < 0) { this.totalPowerConsumption = 0 }
				    	    if (this.voltage1 < 0) { this.voltage1 = 0 }
				    	    if (this.ampere1 < 0) { this.ampere1 = 0 }
					        } else if (this.negative_handling_mode == 1) {	
				             this.powerConsumption = Math.abs(parseFloat(json.emeters[0].power)+parseFloat(json.emeters[1].power)+parseFloat(json.emeters[2].power));
				             this.totalPowerConsumption = Math.abs((parseFloat(json.emeters[0].total)+parseFloat(json.emeters[1].total)+parseFloat(json.emeters[2].total))/1000);
			               this.voltage1 = Math.abs(((parseFloat(json.emeters[0].voltage)+parseFloat(json.emeters[1].voltage)+parseFloat(json.emeters[2].voltage))/3));
				             this.ampere1 = Math.abs(((parseFloat(json.emeters[0].current)*this.pf0)
					            +(parseFloat(json.emeters[1].current)*this.pf1)
					            +(parseFloat(json.emeters[2].current)*this.pf2)));
					          }
							
						}
							
							
					
					if (this.debug_log) { this.log('Successful http response. [ voltage: ' + this.voltage1.toFixed(0) + 'V, current: ' + this.ampere1.toFixed(1) + 'A, consumption: ' + this.powerConsumption.toFixed(0) + 'W, total consumption: ' + this.totalPowerConsumption.toFixed(2) + 'kWh ]'); }
				}
				catch (parseErr) {
					this.log('Error processing data: ' + parseErr.message);
					error = parseErr;
				}
			}
			if (!error) {
				
				resolve(this.powerConsumption,this.totalPowerConsumption,this.voltage1,this.ampere1)
			}
			else {
				reject(error);
			}
			this.waiting_response = false;
		});
	})
	.then((value_current, value_total, value_voltage1, value_ampere1) => {
		if (value_current != null) {
				this.service.getCharacteristic(this._EvePowerConsumption).setValue(value_current, undefined, undefined);
				//FakeGato
				this.historyService.addEntry({time: Math.round(new Date().valueOf() / 1000), power: value_current});}
		if (value_total != null) {
				this.service.getCharacteristic(this._EveTotalConsumption).setValue(value_total, undefined, undefined);}
		if (value_voltage1 != null) {
				this.service.getCharacteristic(this._EveVoltage1).setValue(value_voltage1, undefined, undefined);}
		if (value_ampere1 != null) {
				this.service.getCharacteristic(this._EveAmpere1).setValue(value_ampere1, undefined, undefined);}
		return true;
	}, (error) => {
		return error;
	});
};

EnergyMeter.prototype.getPowerConsumption = function (callback) {
	callback(null, this.powerConsumption);
};

EnergyMeter.prototype.getTotalConsumption = function (callback) {
	callback(null, this.totalPowerConsumption);
};

EnergyMeter.prototype.getVoltage1 = function (callback) {
	callback(null, this.voltage1);
};

EnergyMeter.prototype.getAmpere1 = function (callback) {
	callback(null, this.ampere1);
};

EnergyMeter.prototype.getServices = function () {
	this.log("getServices: " + this.name);
	if (this.update_interval > 0) {
		this.timer = setInterval(this.updateState.bind(this), this.update_interval);
	}
	return [this.informationService, this.service, this.historyService];
};
